cmd/thanos/store.go

```diff
@@ -51,11 +52,23 @@ func registerStore(m map[string]setupFunc, app *kingpin.Application, name string
 	blockSyncConcurrency := cmd.Flag("block-sync-concurrency", "Number of goroutines to use when syncing blocks from object storage.").
 		Default("20").Int()
 
+	minTime := model.TimeOrDuration(cmd.Flag("min-time", "Start of time range limit to serve. Thanos Store serves only blocks, which have start time greater than this value. Option can be a constant time in RFC3339 format or time duration relative to current time, such as -1.5d or 2h45m. Valid duration units are ms, s, m, h, d, w, y.").
```

Comment:
Re 1) `block-min-time` I can name it this way and add `block-max-time`, not a problem.

> (all blocks that overlaps are synced and only exact time range is announced and returned).
> If both max and min are inclusive or both exclusive you use case will not work in all cases. How you will compose chain of store gateways then?

Overlaps are not a problem for us, Thanos Querier will merge responses, so you just make overlaping chain of `--block-min-time` & `--block-max-time`.

> Now if you want to spread blocks evenly among store gateway this gets quite tricky.

This is PR is not about spreading the blocks evenly. I don't really care about that and I don't really see the usecase of spreading blocks evenly.

> How you will partition time when compactor is compacting blocks a lot? When those block time ranges are super dynamic?

Not an issue the sync job already adds new blocks which fit the time range and removes the old ones. So some times it will go into one Thanos Store, sometimes into another.

>I find it extremely not intuitive to work on startTime only. 

We can add 4 flags then: `--block-start-time-min-time`, `--block-start-time-max-time`
, `--block-end-time-min-time`, `--block-end-time-max-time`

> I think @claytono can achieve block paritioning by pure time based partitioning as well -> just make sure time range is not overlapping with not wanted blocks AFAIK

I don't follow, why do we actually care that time range is not overlapping?

The idea is that we need to scale Thanos Store, it's impossible to keep low response times if we don't partition Thanos Store gateway. I.e. what happens to our LRU cache if someone queries for short time period <1w and then a few people on long >3mo, etc. When you shard you can optimize for that type of a deal. Also in general it is more approachable instead of having one big store gateway we have many little ones, which have advantage of easier scheduling (don't need a seperate node for store gw anymore), reducing query latency (every store performs smaller amount of work), etc. 

IMO block juggling is fine. Maybe I need to improve docs more about overlapping case, so that people don't try to do perfect ranges.

Reply:
I agree with @povilasv.  For us this is about horizontal scaling.  We're already partitioning by time by having multiple buckets and creating new ones as needed.  This PR will allow us to reduce a lot of manual work involved in that and automate the rest.

While there may be better solutions to https://github.com/improbable-eng/thanos/issues/448, I think we're going to always have a need for horizontal scaling and we don't have a solution to #448 today.  I definitely want to see the memory efficiency improved, but horizontally scaling has allowed us to ignore the problem for the time being.  This PR gives us breathing room while that's worked out.

Reply:
> and many more clients againts Object store, which as experience shows is faster than having a single client against bucket.

Where you get that from? (: 

Ok @claytono @povilasv so maybe from different position. Let's say we have just `min-time` `max-time` ranges? Nothing block based, but purely time wise. Why is that insufficient? 

My main concern is that tieing some user expierience API (store gw flag) to some hidden format that is **eventually** synced will cause only pain. I would like to avoid this? Having simple time range flags would be both simple, understandable and solve you use case, isn't it?

Reply:
>Where you get that from? (:

For example https://care.qumulo.com/hc/en-us/articles/360015703713-Performance-Results-of-MinIO-with-Qumulo

Take a look at any object store benchmarks, I haven't seen a benchmark which tells that if you use 1 client it's going to be have faster reads then multi client. But you can prove me wrong.

Minio case singe client was:  643 MB/s
Multi-Stream Read: 1.1-1.7GB/s

S3:
Single: 390.8MB/sec
Multi-Stream Read: 1-1.5GB/s

>  Nothing block based, but purely time wise. Why is that insufficient?

Again time duration based is just a simple way to shard thanos store for large deployments.

> My main concern is that tieing some user expierience API (store gw flag) to some hidden format that is eventually synced will cause only pain.

Yes that's the only drawback of this solution, is that people will need to be educated on doing overlapping time ranges, or not doing it at all.

S3 has eventual consistency anyway so in any solution we would have to take similar measures.